src/Microsoft.IdentityModel.Validators/AadIssuerValidator/AadIssuerValidator.cs

```diff
@@ -180,25 +196,30 @@ internal string AadAuthorityV2
 
             try
             {
-                BaseConfigurationManager effectiveConfigurationManager = GetEffectiveConfigurationManager(securityToken);
-                if (validationParameters.RefreshBeforeValidation)
-                    effectiveConfigurationManager.RequestRefresh();
+                var isV2Issuer = IsV2Issuer(securityToken);
+                var effectiveConfigurationManager = GetEffectiveConfigurationManager(isV2Issuer);
 
-                BaseConfiguration configuration = await effectiveConfigurationManager.GetBaseConfigurationAsync(CancellationToken.None).ConfigureAwait(false);
-                string aadIssuer = configuration.Issuer;
-
-                if (!validationParameters.ValidateWithLKG)
+                string aadIssuer = null;
+                if (validationParameters.ValidateWithLKG)
                 {
-                    if (IsValidIssuer(aadIssuer, tenantId, issuer))
-                    {
-                        effectiveConfigurationManager.LastKnownGoodConfiguration = new OpenIdConnectConfiguration() { Issuer = aadIssuer };
-                        return issuer;
-                    }
+                    // returns null if LKG issuer expired
+                    aadIssuer = GetEffectiveLKGIssuer(isV2Issuer);
                 }
                 else
                 {
-                    if (effectiveConfigurationManager.LastKnownGoodConfiguration != null &&
-                        IsValidIssuer(effectiveConfigurationManager.LastKnownGoodConfiguration.Issuer, tenantId, issuer))
+                    var baseConfiguration = await GetBaseConfiguration(effectiveConfigurationManager, validationParameters).ConfigureAwait(false);
+                    aadIssuer = baseConfiguration.Issuer;
+                }
+
+                if (aadIssuer != null)
+                {
+                    var isIssuerValid = IsValidIssuer(aadIssuer, tenantId, issuer);
```

Comment:
would something like this be more clear? 

```cs
var isIssuerValid = IsValidIssuer(aadIssuer, tenantId, issuer);

// The original LKG assignment behavior for previous self-state management.
if (isIssuerValid)
{
    if (!validationParameters.ValidateWithLKG)
        SetEffectiveLKGIssuer(aadIssuer, isV2Issuer, effectiveConfigurationManager.LastKnownGoodLifetime);

    return issuer;
}
```

Reply:
It is another level of nesting, but I find it a little easier to follow. Up to you in the end.